### DISCUSSION TITLE
Lesson 7 Mocks

### DISCUSSION THREAD

**Discussion by MoonChild362000** [Cyfrin/foundry-full-course-cu#882] [Q&A]
I am having an issue where i dont know what is wrong.

this is the issue from the line  - NetworkConfig memory anvilConfig = NetworkConfig({ 

'''
Expected identifier but got 'memory
'''

here's the code 
'''
//SPDX-License-Identifier:MIT

pragma solidity ^0.8.18;

import {Script} from "forge-std/Script.sol";
import {MockV3Aggregator} from "../test/mocks/mockv3aggregator.sol";

contract HelperConfig is Script{

    NetworkConfig public activeNetworkConfig;

    uint8 public constant DECIMALS = 8;
    int256 public constant INITIAL_PRICE = 2000e8;

    struct NetworkConfig{
        address priceFeed;
    }
    
    constructor(){
        if(block.chainid == 11155111){
            activeNetworkConfig = getSepoliaEthConfiq();
        } else {
            activeNetworkConfig = getAnvilEthConfig();
        }
    }

    function getSepoliaEthConfiq() public pure returns(NetworkConfig memory){
        NetworkConfig memory sepoliaConfig = NetworkConfig({
            priceFeed: 0x694AA1769357215DE4FAC081bf1f309aDC325306
        });
        return sepoliaConfig;
    }

     function getAnvilEthConfig()public returns(NetworkConfig memory){
        

        vm.startBroadcast();
        MockV3Aggregator mockPriceFeed = new MockV3Aggregator(
            DECIMALS,
            INITIAL_PRICE
        );
        vm.stopBroadcast();
    }

        NetworkConfig memory anvilConfig = NetworkConfig({ 
            priceFeed: address (mockPriceFeed)
        });
        return anvilConfig;
}
'''

**Comment by 0xusmanf** [comment]
Remove `memory` keyword from both lines where you are declaring `NetworkConfig` variable like I have removed it in the below code.
```solidity
//SPDX-License-Identifier:MIT

pragma solidity ^0.8.18;

import {Script} from "forge-std/Script.sol";
import {MockV3Aggregator} from "../test/mocks/mockv3aggregator.sol";

contract HelperConfig is Script{

NetworkConfig public activeNetworkConfig;

uint8 public constant DECIMALS = 8;
int256 public constant INITIAL_PRICE = 2000e8;

struct NetworkConfig{
    address priceFeed;
}

constructor(){
    if(block.chainid == 11155111){
        activeNetworkConfig = getSepoliaEthConfiq();
    } else {
        activeNetworkConfig = getAnvilEthConfig();
    }
}

function getSepoliaEthConfiq() public pure returns(NetworkConfig memory){
    NetworkConfig sepoliaConfig = NetworkConfig({
        priceFeed: 0x694AA1769357215DE4FAC081bf1f309aDC325306
    });
    return sepoliaConfig;
}

 function getAnvilEthConfig()public returns(NetworkConfig memory){
    if(activeNetworkConfig.priceFeed)

    vm.startBroadcast();
    MockV3Aggregator mockPriceFeed = new MockV3Aggregator(
        DECIMALS,
        INITIAL_PRICE
    );
    vm.stopBroadcast();
}

    NetworkConfig anvilConfig = NetworkConfig({ 
        priceFeed: address (mockPriceFeed)
    });
    return anvilConfig;
}
```

**Reply by MoonChild362000** [reply]
After i deleted it then my return statement for getAnvilEthConfig has a problem which is located at my last line stating that function variable struct or modifier declaration expected.

**Reply by 0xusmanf** [reply]
Directly return the `NetworkConfig` variables instead of declaring and assigning like below.

```solidity
//SPDX-License-Identifier:MIT

pragma solidity ^0.8.18;

import {Script} from "forge-std/Script.sol";
import {MockV3Aggregator} from "../test/mocks/mockv3aggregator.sol";

contract HelperConfig is Script{

NetworkConfig public activeNetworkConfig;

uint8 public constant DECIMALS = 8;
int256 public constant INITIAL_PRICE = 2000e8;

struct NetworkConfig{
    address priceFeed;
}

constructor(){
    if(block.chainid == 11155111){
        activeNetworkConfig = getSepoliaEthConfiq();
    } else {
        activeNetworkConfig = getAnvilEthConfig();
    }
}

function getSepoliaEthConfiq() public pure returns(NetworkConfig memory){
    return NetworkConfig({
        priceFeed: 0x694AA1769357215DE4FAC081bf1f309aDC325306
    });
}

 function getAnvilEthConfig()public returns(NetworkConfig memory){
    if(activeNetworkConfig.priceFeed)

    vm.startBroadcast();
    MockV3Aggregator mockPriceFeed = new MockV3Aggregator(
        DECIMALS,
        INITIAL_PRICE
    );
    vm.stopBroadcast();
}

    return NetworkConfig({ 
        priceFeed: address (mockPriceFeed)
    });
}
```

**Reply by MoonChild362000** [reply]
yo bro by using that i get the same issues too with return

**Reply by Jamster187** [reply]
yo MoonChild, you have closed your getAnvilEthConfig() function too early in your code.

In your "getAnvilEthConfig()" function, after "vm.stopBroadcast();" move the "}" to after the

```
return NetworkConfig({ 
        priceFeed: address (mockPriceFeed)
    });
```
part of your code. Keep the memory keyword in both your anvil and sepolia eth config functions and your code should work.

For reference, your code should look like this after the modification:

```
//SPDX-License-Identifier:MIT

pragma solidity ^0.8.18;

import {Script} from "forge-std/Script.sol";
import {MockV3Aggregator} from "../test/mocks/mockv3aggregator.sol";

contract HelperConfig is Script {NetworkConfig public activeNetworkConfig;

uint8 public constant DECIMALS = 8;
int256 public constant INITIAL_PRICE = 2000e8;

struct NetworkConfig{
    address priceFeed;
}

constructor(){
    if(block.chainid == 11155111){
        activeNetworkConfig = getSepoliaEthConfiq();
    } else {
        activeNetworkConfig = getAnvilEthConfig();
    }
}

function getSepoliaEthConfiq() public pure returns(NetworkConfig memory){
    NetworkConfig memory sepoliaConfig = NetworkConfig({
        priceFeed: 0x694AA1769357215DE4FAC081bf1f309aDC325306
    });
    return sepoliaConfig;
}

function getAnvilEthConfig()public returns(NetworkConfig memory){
    
    vm.startBroadcast();
    MockV3Aggregator mockPriceFeed = new MockV3Aggregator(
        DECIMALS,
        INITIAL_PRICE
    );
    vm.stopBroadcast();


    NetworkConfig memory anvilConfig = NetworkConfig({ 
        priceFeed: address (mockPriceFeed)
    });
    return anvilConfig;
}
}
```

Also, I think it's worth mentioning that you spelled "config" with a q for your "getSepoliaEthConfiq()" function - this may result in some unnecessary debugging if you forget that detail about your code as you continue to develop it. GL with the rest of the course!
